### PR TITLE
Clients: show did length estimation in downloadclient. Closes #3824

### LINKED ISSUE #3824: Incomplete download looks "complete" when restricting RSEs
Motivation
----------
As a user one may naively assume that a dataset/container is completely downloaded when seeing an output like
```
Total files :                                93
Downloaded files :                            0
Files already found locally :                93
Files that cannot be downloaded :             0
```
and a return code of 0. However, this example is taken from a case where I restricted the RSEs to download from, e.g., via `rucio download --rse '*\SOME_PROBLEMATIC_SITE' someContainer` and the full container contains more than the 93 files reported. (Downloads from the problematic site in this case keep failing for a subset of the files involved)

The effect is the same if some RSE is explicitly selected (`--rse 'SOME_GOOD_RSE'`) or individual RSEs are vetoed as in the example above.

Modification
------------
The total number of files should likely be determined without the RSE restriction to avoid files only present on the excluded RSEs being forgotten about. If the current counting is intentional it would be good to at least print a warning in case only a restricted set of RSEs was considered.

### PR DESCRIPTION
**Pull request by rcarpa**
Some refactoring was needed to make it possible. see individual commits for details. 
I would keep the commits separated on merge instead of squashing them.

### CODE REVIEW

File: lib/rucio/tests/test_download.py [hunk 1]
```diff
@@ -409,26 +512,45 @@ def test_trace_copy_out_and_checksum_validation(vo, rse_factory, did_factory, do
...
+
+    with TemporaryDirectory() as tmp_dir:
         nrandom = 2
-        result = download_client.download_dids([{'did': dataset_did_str, 'nrandom': nrandom, 'base_dir': tmp_dir}])
+        result = download_client.download_dids([{'did': dataset1_did_str, 'nrandom': nrandom, 'base_dir': tmp_dir}])
         assert len(result) == nrandom
 
 
-def test_download_blocklisted_replicas(rse_factory, did_factory, download_client):
+def test_download_blacklisted_replicas(rse_factory, did_factory, download_client):
```
Comment (mlassnig) [inline lib/rucio/tests/test_download.py]: blocklist was quite alright :-)

File: bin/rucio [hunk 2]
```diff
@@ -1090,30 +1090,33 @@ def download(args):
...
+        print('DID %s' % summary_key)
+        length = did_summary['length']
+        ds_total = did_summary['_total']
+        downloaded_files = did_summary['DONE']
+        local_files = did_summary['ALREADY_DONE']
         not_downloaded_files = ds_total - downloaded_files - local_files
-        print('{0:40} {1:6d}'.format('Total files: ', ds_total))
+
+        if length:
+            print('{0:40} {1:6d}'.format('Total files (did): ', length))
```
Comment (mlassnig) [inline bin/rucio]: I would uppercase `DID`

File: bin/rucio [hunk 3]
```diff
@@ -1090,30 +1090,33 @@ def download(args):
...
+            print('{0:40} {1:6d}'.format('Total files (after filter):   ', ds_total))
```
Comment (mlassnig) [inline bin/rucio]: shorter: `filtered`

**Comment by rcarpa** [comment]
I cannot understand why that multi-vo tests fails after adding this function. But it's consistent :/

**Comment by bari12** [comment]
@Thysk any idea what is going on with the Multi-VO tests here? It is a very strange behaviour @rcarpa is noticing. Basically the code he changed does not touch the failing test at all. But it is working on master...

**Comment by thysk** [comment]
@bari12 It looks like its succeeding in the 1st VO, but then is failing in the second VO
But given its the same process each time not really sure why
If this something that is failing in other PRs as well?
As there are multiple tests after this that function perfectly well, and perform almost identical tasks
Is there a way to see the full logs of the tests? because after the assert that fails there is another assert and I would be curious to see if that had the correct number of files (though it looks from the bytes that no files have been added to the dataset)

Given that all the exact same processes are used in the next test (test_account_counter_judge_evaluate_detach)
In am not sure how it is only failing the one test. And only on this PR, as, as you said, the PR doesnt touch any of the code that runs in the tests (as far as I can tell)

**Comment by bari12** [comment]
@rcarpa Please rebase and run again.
I don't think this is failing in other tests...

**Comment by rcarpa** [comment]
I rebased and the problem persists :confused:

**Review by mlassnig — APPROVED** [review]

**Comment by bari12** [comment]
Thanks @rcarpa 
I'll bring this in as a feature for 1.27 though, since I want to stabilise the 1.26 clients a bit at the moment and not introduce yet another change.

**Pull request merged**